pkg/securitypolicy/framework.rego

```diff
@@ -0,0 +1,115 @@
+package framework
+
+import future.keywords.every
+import future.keywords.in
+
+default mount_device := false
+mount_device := true {
+    some container in data.policy.containers
+    some layer in container.layers
+    input.deviceHash == layer
+}
+
+layerPaths_ok(container) {
+    length := count(container.layers)
+    count(input.layerPaths) == length
+    every i, path in input.layerPaths {
+        container.layers[length - i - 1] == data.devices[path]
+    }
+}
+
+default mount_overlay := false
+mount_overlay := true {
+    some container in data.policy.containers
+    layerPaths_ok(container)
+}
+
+command_ok(container) {
+    count(input.argList) == count(container.command)
+    every i, arg in input.argList {
+        container.command[i] == arg
+    }
+}
+
+env_ok(pattern, "string", value) {
+    pattern == value
+}
+
+env_ok(pattern, "re2", value) {
+    regex.match(pattern, value)
+}
+
+envList_ok(container) {
+    every env in input.envList {
+        some rule in container.env_rules
+        env_ok(rule.pattern, rule.strategy, env)
+    }
+}
+
+workingDirectory_ok(container) {
+    input.workingDir == container.working_dir
+}
+
+default create_container := false
+create_container := true {
+    not input.containerID in data.started
+    some container in data.policy.containers
+    layerPaths_ok(container)
+    command_ok(container)
+    envList_ok(container)
+    workingDirectory_ok(container)
+    mountList_ok(container)
+}
+
+mountSource_ok(constraint, source) {
+    startswith(constraint, data.sandboxPrefix)
+    newConstraint := replace(constraint, data.sandboxPrefix, input.sandboxDir)
+    regex.match(newConstraint, source)
+}
+
+mountSource_ok(constraint, source) {
+    startswith(constraint, data.hugePagesPrefix)
+    newConstraint := replace(constraint, data.hugePagesPrefix, input.hugePagesDir)
+    regex.match(newConstraint, source)
+}
+
+mountSource_ok(constraint, source) {
+    constraint == source
+}
+
+mountConstraint_ok(constraint, mount) {
+    mount.type == constraint.type
+    mountSource_ok(constraint.source, mount.source)
+    mount.destination != ""
+    mount.destination == constraint.destination
+    every option in mount.options {
```

Comment:
are we missing `count(mount.optinos) == count(constraint.options)`?

Reply:
No harm in adding it (as it will short-circuit the check which follows) but by checking for a bijective mapping the count being equal is proven as well.

Reply:
As you can see I've added in the count check because it acts as a small optimization of this query, but the fact that it isn't strictly needed is an important point re: how Rego works. If it isn't clear, please feel free to reach out to me or @SeanTAllen and we can walk you through it 😊